### PR TITLE
KP-7610 Using SEQ attribute to determine page download URLs

### PR DESCRIPTION
**Pull request by aajarven**
TL;DR:
- download URLs now based on page number, derived from SEQ attribute of   access image elements
- ALTOs created by creating an ALTO counterpart for each access image
- File naming changed: we now have 00001.xml and 00001.jp2 as opposed to  00001.xml and pr-00001.jp2
- No longer verifying that there are no unexpected file groups (none were found during the [KP-7716](https://jira.eduuni.fi/browse/KP-7716) run

Previously the download URLs for both ALTO and access image files were determined based on the file name in the METS file: e.g. image with file name "pr-00003.jp2" would get a download URL ending "/image/3". Unfortunately this does not always work, as there can be other files (target images etc) listed in METS with lower-numbered file names, or even file names between the actual pages. The digi.kansalliskirjasto.fi download URLs do not follow the file names though: the scanned image for the first page is always "/image/1" and the ALTO file for it "page-1.xml" and so on.

For access images, it's easy to fix this discrepancy: each of their file elements in METS has a SEQ attribute that corresponds to the page number. For ALTOs, this is not as simple: they don't have SEQ attribute and the same file group contais OCRs for both the actual page scans and other files such as target images.

Thus, the METS parsing was updated so that first all access image files are extracted, and then an ALTO file is generated for each of them. This means that no information directly from the file elements for ALTO files is actually used. If such functionality was desired (e.g. if we start verifying download checksums), it would likely require matching the ALTO and access image elements by comparing the numeric part of the file name. No such matching was implemented at this stage though.

The classes for ALTO and access image files also diverged a bit due to this change: only access images take the location_xlink when initializing. Instead, all files know their page number.

Only inspecting the fileGrp for access images means that we no longer have verification that the METS does not contain surprising file group USE and ID pairs that we should process. The earlier DAG runs have not uncovered such file groups though.

Against all best practices, the file naming convention was renewed at the same time, so that all files have a name that consists of the page number (padded with leading zeroes so that it is 5 digits long) and a file type extension.

### CODE REVIEW

File: harvester/mets.py [hunk 1]
```diff
@@ -61,13 +61,59 @@ def _ensure_files(self):
...
+            parent = file_element.getparent()
+
+            filegrp_use = parent.attrib["USE"]
+            filegrp_id = parent.attrib["ID"]
+
+            if filegrp_use in ["Images", "reference"] and filegrp_id in [
+                "IMGGRP",
+                "ACIMGGRP",
+            ]:
+                page_number = int(file_element.attrib["SEQ"])
```
Comment (Traubert) [inline harvester/mets.py:88]: If this raises an exception, do we ultimately get a failed task? I think so, via `_ensure_files()` -> `files()` -> `SavePageFilesSFTPOperator.execute()` -> `SaveAltosSFTPOperator.execute()` -> `StowBindingBatchOperator()`. What I mean is, are we so sure of these always existing that we don't want to check?

**Review by Traubert — APPROVED** [review]

File: harvester/mets.py (hunk 1 above)
Comment (aajarven) [reply harvester/mets.py:88]: Yea, the task will now fail if we can't find the page number.

NLF gave us a soft promise that all the METSes currently in the collection should have them, and at the time we don't really have a good alternative for figuring out the mapping between elements that isn't complex (mapping the file elements to other elements in other sections of the METS via their identifiers and deducing the URLs based on that) and still uncertain (we still don't know if all that metadata is consistent and always available). So even if we checked, at the moment we wouldn't have a good alternative to just manually raising an exception.

If it turns out that SEQs are still missing in the new collection, then we need to figure out what those bindings are like and how we can work with them, but that requires actually seeing what metadata they actually have.
